src/ipyradiant/visualization/interactive_exploration.py

```diff
@@ -0,0 +1,324 @@
+            self.cyto_graph,
+            W.HBox(
+                children=[
+                    self.expand_button,
+                    self.undo_button,
+                    self.remove_temp_nodes_button,
+                ]
+            ),
+        )
+
+    @trt.validate("children")
+    def validate_children(self, proposal):
+        children = proposal.value
+        if not children:
+            children = (
+                self.cyto_graph,
+                W.HBox(
+                    children=[
+                        self.expand_button,
+                        self.undo_button,
+                        self.remove_temp_nodes_button,
+                    ]
+                ),
+            )
+        return children
+
+    def get_node(self, node):
+        """
+        This function is used to find a node given the id of a node copy. Used in the log_node_clicks
+        method to change the color of nodes.
+        """
+
+        for node_obj in self.cyto_graph.graph.nodes:
+            if node_obj.data["id"] == node["data"]["id"]:
+                return node_obj
+        # maybe return None and log warning?
+        raise ValueError("Node not found in cytoscape.graph.nodes.")
+
+    def log_node_clicks(self, node):
+        """
+        This function works with registering a click on a node. This will mark the node as selected and change the color of the
+        selected node.
+        """
+
+        try:
+            node_object = self.get_node(node)
+        except ValueError:
+            #     # logger.warn
+            print("Node {} not found in cytoscape graph.".format(node["data"]["id"]))
+            return
+
+        if self.selected_node == node_object:
+            node_object.classes = "clicked"
+            self.cyto_graph.graph.add_node(Node(data={"id": "random node"}))
+            self.cyto_graph.graph.remove_node_by_id("random node")
+            # NOTE: Class changes won't propogate to the front end for added nodes until
+            # the graph is updated.
+            # To fix this we create a random node and then quickly delete it so that the changes propogate.
+            # TODO: Add logger.warning to signal this event
+
+        self.selected_node = node_object
+
+    def expand_button_clicked(self, b):
+        """
+        This function expands a node by loading in its predicates and subjects when
+        a node is selected and the expand button is clicked.
+        """
+        self.undo_button.disabled = False
+        if self.selected_node is None:
+            return None
+        new_data = GetOutgoingPredicateObjects.run_query(
+            graph=self.rdf_graph, subject=self.selected_node.data["iri"]
+        )
+        objs = new_data["o"].tolist()
+        preds = new_data["p"].tolist()
+        labels = new_data["label"].tolist()
+        # add nodes
+        self.existing_node_ids = [
+            node.data["id"] for node in self.cyto_graph.graph.nodes
+        ]
+        self.new_nodes = {}
```

Comment:
This is begging for a dictionary comprehension...  Something like this:

(also made the variable names a bit more explicit in the comprehension to help with readability)

```python
self.new_nodes = {
    Node(
        data={
            "id": str(object_iri),
            "iri": object_iri,
            "_label": labels[object_idx] or str(object_iri),
        },
        classes="temp",
    )
    for object_idx, object_iri in enumerate(objs)
    if str(object_iri) not in self.existing_node_ids
}

self.cyto_graph.graph.add_nodes_from(self.new_nodes)
```

Reply:
This presumes that `self.cyto_graph.graph` is a networkx graph.